File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        nodeDict = dict()
        parent = dict()

        for p, c, d in descriptions:
            if p not in nodeDict:
                nodeDict[p] = TreeNode(p)
            if c not in nodeDict:
                nodeDict[c] = TreeNode(c)
            if d: nodeDict[p].left = nodeDict[c]
            else: nodeDict[p].right = nodeDict[c]
            if p not in parent: parent[p] = -1
            parent[c] = p

        for key, value in parent.items():
            if value == -1:
                return nodeDict[key]

        return

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.java
class Solution {
    public TreeNode createBinaryTree(int[][] descriptions) {
        Map<Integer, TreeNode> nodeDict = new HashMap<>();
        Map<Integer, Integer> parent = new HashMap<>();

        for (int i = 0; i < descriptions.length; i++) {
            if (!nodeDict.containsKey(descriptions[i][0])) {
                nodeDict.put(descriptions[i][0], new TreeNode(descriptions[i][0]));
            }
            if (!nodeDict.containsKey(descriptions[i][1])) {
                nodeDict.put(descriptions[i][1], new TreeNode(descriptions[i][1]));
            }
            if (descriptions[i][2] == 1) {
                nodeDict.get(descriptions[i][0]).left = nodeDict.get(descriptions[i][1]);
            } else {
                nodeDict.get(descriptions[i][0]).right = nodeDict.get(descriptions[i][1]);
            }
            if (!parent.containsKey(descriptions[i][0])) {
                parent.put(descriptions[i][0], -1);
            }
            parent.put(descriptions[i][1], descriptions[i][0]);
        }

        for (Map.Entry<Integer, Integer> map: parent.entrySet()) {
            int key = map.getKey();
            if (parent.get(key) == -1) {
                return nodeDict.get(key);
            }
        }

        return null;
    }
}